File: Clients/Node/Core/src/signalr/TaskHubClient.ts
import type * as signalR from '@microsoft/signalr';
import { BaseSignalRConnection } from './BaseSignalRConnection';
import { 
  SignalREndpoints,
  type ITaskHubServer,
  type TaskStartedEvent,
  type TaskProgressEvent,
  type TaskCompletedEvent,
  type TaskFailedEvent,
  type TaskCancelledEvent,
  type TaskTimedOutEvent
} from '../models/signalr';

/**
 * SignalR client for the Task Hub, providing real-time task progress notifications.
 */
export class TaskHubClient extends BaseSignalRConnection implements ITaskHubServer {
  /**
   * Gets the hub path for task notifications.
   */
  protected get hubPath(): string {
    return SignalREndpoints.TaskHub;
  }

  /**
   * Event handlers for task notifications.
   */
  onTaskStarted?: (event: TaskStartedEvent) => Promise<void>;
  onTaskProgress?: (event: TaskProgressEvent) => Promise<void>;
  onTaskCompleted?: (event: TaskCompletedEvent) => Promise<void>;
  onTaskFailed?: (event: TaskFailedEvent) => Promise<void>;
  onTaskCancelled?: (event: TaskCancelledEvent) => Promise<void>;
  onTaskTimedOut?: (event: TaskTimedOutEvent) => Promise<void>;

  /**
   * Configures the hub-specific event handlers.
   */
  protected configureHubHandlers(connection: signalR.HubConnection): void {
    connection.on('TaskStarted', async (taskId: string, taskType: string, metadata: unknown) => {
      console.warn(`Task started: ${taskId}, Type: ${taskType}`);
      if (this.onTaskStarted) {
        await this.onTaskStarted({ eventType: 'TaskStarted', taskId, taskType, metadata: metadata as Record<string, unknown> });
      }
    });

    connection.on('TaskProgress', async (taskId: string, progress: number, message?: string) => {
      console.warn(`Task progress: ${taskId}, Progress: ${progress}%`);
      if (this.onTaskProgress) {
        await this.onTaskProgress({ eventType: 'TaskProgress', taskId, progress, message });
      }
    });

    connection.on('TaskCompleted', async (taskId: string, result: unknown) => {
      console.warn(`Task completed: ${taskId}`);
      if (this.onTaskCompleted) {
        await this.onTaskCompleted({ eventType: 'TaskCompleted', taskId, result: result as Record<string, unknown> });
      }
    });

    connection.on('TaskFailed', async (taskId: string, error: string, isRetryable: boolean) => {
      console.error(`Task failed: ${taskId}, Error: ${error}, Retryable: ${isRetryable}`);
      if (this.onTaskFailed) {
        await this.onTaskFailed({ eventType: 'TaskFailed', taskId, error, isRetryable });
      }
    });

    connection.on('TaskCancelled', async (taskId: string, reason?: string) => {
      console.warn(`Task cancelled: ${taskId}, Reason: ${reason}`);
      if (this.onTaskCancelled) {
        await this.onTaskCancelled({ eventType: 'TaskCancelled', taskId, reason });
      }
    });

    connection.on('TaskTimedOut', async (taskId: string, timeoutSeconds: number) => {
      console.error(`Task timed out: ${taskId}, Timeout: ${timeoutSeconds}s`);
      if (this.onTaskTimedOut) {
        await this.onTaskTimedOut({ eventType: 'TaskTimedOut', taskId, timeoutSeconds });
      }
    });
  }

  /**
   * Subscribe to notifications for a specific task.
   */
  async subscribeToTask(taskId: string): Promise<void> {
    if (!taskId?.trim()) {
      throw new Error('Task ID cannot be null or empty');
    }

    await this.invoke('SubscribeToTask', taskId);
    console.warn(`Subscribed to task: ${taskId}`);
  }

  /**
   * Unsubscribe from notifications for a specific task.
   */
  async unsubscribeFromTask(taskId: string): Promise<void> {
    if (!taskId?.trim()) {
      throw new Error('Task ID cannot be null or empty');
    }

    await this.invoke('UnsubscribeFromTask', taskId);
    console.warn(`Unsubscribed from task: ${taskId}`);
  }

  /**
   * Subscribe to notifications for all tasks of a specific type.
   */
  async subscribeToTaskType(taskType: string): Promise<void> {
    if (!taskType?.trim()) {
      throw new Error('Task type cannot be null or empty');
    }

    await this.invoke('SubscribeToTaskType', taskType);
    console.warn(`Subscribed to task type: ${taskType}`);
  }

  /**
   * Unsubscribe from notifications for a task type.
   */
  async unsubscribeFromTaskType(taskType: string): Promise<void> {
    if (!taskType?.trim()) {
      throw new Error('Task type cannot be null or empty');
    }

    await this.invoke('UnsubscribeFromTaskType', taskType);
    console.warn(`Unsubscribed from task type: ${taskType}`);
  }

  /**
   * Subscribe to multiple tasks at once.
   */
  async subscribeToTasks(taskIds: string[]): Promise<void> {
    await Promise.all(taskIds.map(taskId => this.subscribeToTask(taskId)));
  }

  /**
   * Unsubscribe from multiple tasks at once.
   */
  async unsubscribeFromTasks(taskIds: string[]): Promise<void> {
    await Promise.all(taskIds.map(taskId => this.unsubscribeFromTask(taskId)));
  }

  /**
   * Subscribe to multiple task types at once.
   */
  async subscribeToTaskTypes(taskTypes: string[]): Promise<void> {
    await Promise.all(taskTypes.map(taskType => this.subscribeToTaskType(taskType)));
  }

  /**
   * Unsubscribe from multiple task types at once.
   */
  async unsubscribeFromTaskTypes(taskTypes: string[]): Promise<void> {
    await Promise.all(taskTypes.map(taskType => this.unsubscribeFromTaskType(taskType)));
  }
}

File: archive/Clients.DotNet/ConduitLLM.CoreClient/src/SignalR/TaskHubClient.cs
using Microsoft.AspNetCore.SignalR.Client;
using Microsoft.Extensions.Logging;
using ConduitLLM.CoreClient.Models;

namespace ConduitLLM.CoreClient.SignalR;

/// <summary>
/// SignalR client for the Task Hub, providing real-time task progress notifications.
/// </summary>
public class TaskHubClient : BaseSignalRConnection, ITaskHubServer
{
    /// <summary>
    /// Initializes a new instance of the TaskHubClient class.
    /// </summary>
    /// <param name="baseUrl">Base URL of the Conduit Core API.</param>
    /// <param name="virtualKey">Virtual key for authentication.</param>
    /// <param name="logger">Optional logger instance.</param>
    public TaskHubClient(string baseUrl, string virtualKey, ILogger<TaskHubClient>? logger = null)
        : base(baseUrl, virtualKey, logger)
    {
    }

    /// <summary>
    /// Gets the hub path for task notifications.
    /// </summary>
    protected override string HubPath => SignalREndpoints.TaskHub;

    #region Events

    /// <summary>
    /// Event raised when a task is started.
    /// </summary>
    public event Func<string, string, object, Task>? TaskStarted;

    /// <summary>
    /// Event raised when task progress is updated.
    /// </summary>
    public event Func<string, int, string?, Task>? TaskProgress;

    /// <summary>
    /// Event raised when a task completes successfully.
    /// </summary>
    public event Func<string, object, Task>? TaskCompleted;

    /// <summary>
    /// Event raised when a task fails.
    /// </summary>
    public event Func<string, string, bool, Task>? TaskFailed;

    /// <summary>
    /// Event raised when a task is cancelled.
    /// </summary>
    public event Func<string, string?, Task>? TaskCancelled;

    /// <summary>
    /// Event raised when a task times out.
    /// </summary>
    public event Func<string, int, Task>? TaskTimedOut;

    #endregion

    /// <summary>
    /// Configures the hub-specific event handlers.
    /// </summary>
    /// <param name="connection">The hub connection to configure.</param>
    protected override void ConfigureHubHandlers(HubConnection connection)
    {
        connection.On<string, string, object>("TaskStarted", async (taskId, taskType, metadata) =>
        {
            _logger?.LogDebug("Task started: {TaskId}, Type: {TaskType}", taskId, taskType);
            if (TaskStarted != null)
            {
                await TaskStarted.Invoke(taskId, taskType, metadata);
            }
        });

        connection.On<string, int, string?>("TaskProgress", async (taskId, progress, message) =>
        {
            _logger?.LogDebug("Task progress: {TaskId}, Progress: {Progress}%", taskId, progress);
            if (TaskProgress != null)
            {
                await TaskProgress.Invoke(taskId, progress, message);
            }
        });

        connection.On<string, object>("TaskCompleted", async (taskId, result) =>
        {
            _logger?.LogDebug("Task completed: {TaskId}", taskId);
            if (TaskCompleted != null)
            {
                await TaskCompleted.Invoke(taskId, result);
            }
        });

        connection.On<string, string, bool>("TaskFailed", async (taskId, error, isRetryable) =>
        {
            _logger?.LogDebug("Task failed: {TaskId}, Error: {Error}, Retryable: {Retryable}", 
                taskId, error, isRetryable);
            if (TaskFailed != null)
            {
                await TaskFailed.Invoke(taskId, error, isRetryable);
            }
        });

        connection.On<string, string?>("TaskCancelled", async (taskId, reason) =>
        {
            _logger?.LogDebug("Task cancelled: {TaskId}, Reason: {Reason}", taskId, reason);
            if (TaskCancelled != null)
            {
                await TaskCancelled.Invoke(taskId, reason);
            }
        });

        connection.On<string, int>("TaskTimedOut", async (taskId, timeoutSeconds) =>
        {
            _logger?.LogDebug("Task timed out: {TaskId}, Timeout: {Timeout}s", taskId, timeoutSeconds);
            if (TaskTimedOut != null)
            {
                await TaskTimedOut.Invoke(taskId, timeoutSeconds);
            }
        });
    }

    #region ITaskHubServer Implementation

    /// <summary>
    /// Subscribe to notifications for a specific task.
    /// </summary>
    /// <param name="taskId">Task identifier to subscribe to.</param>
    public async Task SubscribeToTask(string taskId)
    {
        if (string.IsNullOrWhiteSpace(taskId))
            throw new ArgumentException("Task ID cannot be null or empty", nameof(taskId));

        await InvokeAsync("SubscribeToTask", new object[] { taskId });
        _logger?.LogDebug("Subscribed to task: {TaskId}", taskId);
    }

    /// <summary>
    /// Unsubscribe from notifications for a specific task.
    /// </summary>
    /// <param name="taskId">Task identifier to unsubscribe from.</param>
    public async Task UnsubscribeFromTask(string taskId)
    {
        if (string.IsNullOrWhiteSpace(taskId))
            throw new ArgumentException("Task ID cannot be null or empty", nameof(taskId));

        await InvokeAsync("UnsubscribeFromTask", new object[] { taskId });
        _logger?.LogDebug("Unsubscribed from task: {TaskId}", taskId);
    }

    /// <summary>
    /// Subscribe to notifications for all tasks of a specific type.
    /// </summary>
    /// <param name="taskType">Task type to subscribe to.</param>
    public async Task SubscribeToTaskType(string taskType)
    {
        if (string.IsNullOrWhiteSpace(taskType))
            throw new ArgumentException("Task type cannot be null or empty", nameof(taskType));

        await InvokeAsync("SubscribeToTaskType", new object[] { taskType });
        _logger?.LogDebug("Subscribed to task type: {TaskType}", taskType);
    }

    /// <summary>
    /// Unsubscribe from notifications for a task type.
    /// </summary>
    /// <param name="taskType">Task type to unsubscribe from.</param>
    public async Task UnsubscribeFromTaskType(string taskType)
    {
        if (string.IsNullOrWhiteSpace(taskType))
            throw new ArgumentException("Task type cannot be null or empty", nameof(taskType));

        await InvokeAsync("UnsubscribeFromTaskType", new object[] { taskType });
        _logger?.LogDebug("Unsubscribed from task type: {TaskType}", taskType);
    }

    #endregion

    #region Convenience Methods

    /// <summary>
    /// Subscribe to multiple tasks at once.
    /// </summary>
    /// <param name="taskIds">Task identifiers to subscribe to.</param>
    public async Task SubscribeToTasksAsync(IEnumerable<string> taskIds)
    {
        var tasks = taskIds.Select(SubscribeToTask);
        await Task.WhenAll(tasks);
    }

    /// <summary>
    /// Unsubscribe from multiple tasks at once.
    /// </summary>
    /// <param name="taskIds">Task identifiers to unsubscribe from.</param>
    public async Task UnsubscribeFromTasksAsync(IEnumerable<string> taskIds)
    {
        var tasks = taskIds.Select(UnsubscribeFromTask);
        await Task.WhenAll(tasks);
    }

    /// <summary>
    /// Subscribe to multiple task types at once.
    /// </summary>
    /// <param name="taskTypes">Task types to subscribe to.</param>
    public async Task SubscribeToTaskTypesAsync(IEnumerable<string> taskTypes)
    {
        var tasks = taskTypes.Select(SubscribeToTaskType);
        await Task.WhenAll(tasks);
    }

    /// <summary>
    /// Unsubscribe from multiple task types at once.
    /// </summary>
    /// <param name="taskTypes">Task types to unsubscribe from.</param>
    public async Task UnsubscribeFromTaskTypesAsync(IEnumerable<string> taskTypes)
    {
        var tasks = taskTypes.Select(UnsubscribeFromTaskType);
        await Task.WhenAll(tasks);
    }

    #endregion
}